File: problems/easy/1_two_sum/solution.py
"""
Time Complexity: O(N) – In the worst case, the for-loop iterates through all N elements, performing O(1) operations (lookup and insertion) per iteration on average.

Space Complexity: O(N) – In the worst case, the hash map stores all N elements of nums before a valid pair is found.
"""
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:

        # Create a hash map to store number-to-index mapping
        hashmap = dict()

        # Iterate through the array with enumeration to get both index and value
        for index, num in enumerate(nums):

            # Calculate the complement
            complement = target - num
            
            # Check if the complement exists in the map
            if complement in hashmap:

                # Return the indices if complement is found
                return [index, hashmap[complement]]

            # Insert the current number and its index into the map
            hashmap[num] = index

File: problems/easy/1_two_sum/solution.rs
"""
Time Complexity: O(N)
Space Complexity: O(N)
"""
use std::collections::HashMap;

impl Solution {
    pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {

        // Create a hash map to store number-to-index mapping
        let mut num_map = HashMap::with_capacity(nums.len());
        
        // Iterate through the array with enumeration to get both index and value
        for (index, &num) in nums.iter().enumerate() {

            // Calculate the complement
            let complement = target - num;
            
            // Check if the complement exists in the map
            if let Some(&complement_index) = num_map.get(&complement) {

                // Return the indices if complement is found
                return vec![index as i32, complement_index as i32];
            }
            
            // Insert the current number and its index into the map
            num_map.insert(num, index);
        }

        // This should never happen given the problem constraints
        unreachable!()
    }
}
